Container docs rewrite and reorganization (#14527)
microsoft/FluidFramework: docs/content/docs/build/container-states-events.md

Clients could continue to edit the container's data, but the data isn't shared anymore. Each client's changes are known only to it. 

##### Deleting a container from the service

Deleting a container is a service-specific feature, so you should refer to the documentation associated with the specific Fluid service you are using for instructions on how to do it. 
To prevent clients from being suddenly cut off from data sharing, the service may put restrictions on when a container can be deleted. 
For example, it may allow deletion only if all clients have been disconnected from the container for a specified period of time. 

Once it is deleted, the container cannot move back to the **published** state again.
It is possible for a client to create a new container using the same schema as a deleted container. 
If a record of the final data values of the deleted container are available to the creating client, they can be used to initialize the new container. 

> But the new container has a different ID from the deleted one and subsequent clients will have to connect to it using the new ID.

Author: @alexvy86 We refer to the ID returned from `container.attach()` and the one passed to `getContainer()` as the container ID all over our docs, including the Getting Started docs. I suggest that we continue this because introducing the concept of "document" would be confusing. When the ID of the local `container` object is what's being referred to, let's call it "local container object ID" and, where helpful, not parenthetically that it is not the same as the container ID. How does that sound to you?

> But the new container has a different ID from the deleted one and subsequent clients will have to connect to it using the new ID.

Reviewer: I think that's alright for now to avoid over-complicating the docs but would still look for a place to put a callout note close to the "start" of the docs (not necessarily the start of this file, although it might be a good place for it too), to point out the fact that the terminology used here does not necessarily match one-to-one what they'd see when debugging/troubleshooting.

> But the new container has a different ID from the deleted one and subsequent clients will have to connect to it using the new ID.

Author: @alexvy86 
I just noticed that our reference docs for the `FluidContainer` class and the `IFluidContainer` interface indicate that they do not have an `id` property. So, what is this "containerId" you are referring to that is distinct from the "documentId"?

> But the new container has a different ID from the deleted one and subsequent clients will have to connect to it using the new ID.

Reviewer: Ah... ok, we probably want to figure out which part of our API these docs are meant for. `IFluidContainer/FluidContainer` are the main entry points for users using the Azure client (probably most 3rd party devs), but they abstract away many things, so maybe the feedback I've been giving doesn't apply in that case. @ssimic2 is probably more familiar with that area. Sasha, as an initial question, how strong is the distinction between containerId and documentId in the Azure client API? Seems like it might be completely hidden from the users?

#### Handling publication status

Your code can test for the publication status with the [container.AttachState]({{< relref "/docs/apis/fluid-static/fluidcontainer-class.md#attachstate-property" >}}) property which has an [AttachState]({{< relref "/docs/apis/container-definitions.md#attachstate-enum" >}}) value. 
This can be useful if your application will publish the container in some code paths, but not others. 
For example, in complex cases, you may need to determine whether the container has already been published before you call `container.attach`. The following is a simple example.

```typescript
if (container.attachState !== AttachState.Attached) {
    container.attach();
}

> ```

Reviewer: I don't think this is the best example. Two clients with detached/unpublished containers will always be able to separately call `attach()` on their containers and succeed, but that will result in two separate backing documents in the service. Before a container is attached, there is no concept of several clients sharing it. The APIs are also designed in a way where this scenario doesn't arise. If a client wants to create a new container, they'll use the "create new" flow (`loader.createDetachedContainer()`), and if a client already has a "container" (document) id, it'll use the "load existing flow" (`loader.loadExisting(id)`).

A better example might be logic that should only run in the creating client *before* a container is attached/published. Every client that loads that "container" (the document that backs it) might have to go through a code path like this:

```typescript
if (container.attachState !== AttachState.Attached) {
  // First-time initialization that only runs in the creating client
}
```

In all subsequent containers, `container.attachState === AttachState.Attached` so they'll skip that code.

Note this also means that subsequent clients will never see `AttachState.Detached` nor `AttachState.Attaching`.

Author: > I don't think this is the best example. Two clients with detached/unpublished containers will always be able to separately call `attach()` on their containers and succeed, but that will result in two separate backing documents in the service. Before a container is attached, there is no concept of several clients sharing it. The APIs are also designed in a way where this scenario doesn't arise. If a client wants to create a new container, they'll use the "create new" flow (`loader.createDetachedContainer()`), and if a client already has a "container" (document) id, it'll use the "load existing flow" (`loader.loadExisting(id)`).
> 
> A better example might be logic that should only run in the creating client _before_ a container is attached/published. Every client that loads that "container" (the document that backs it) might have to go through a code path like this:
> 
> ```ts
> if (container.attachState !== AttachState.Attached) {
>   // First-time initialization that only runs in the creating client
> }
> ```
> 
> In all subsequent containers, `container.attachState === AttachState.Attached` so they'll skip that code.
> 
> Note this also means that subsequent clients will never see `AttachState.Detached` nor `AttachState.Attaching`.

@alexvy86 Actually, your suggestion is exactly what I intended. I'll try to clarify this passage.

How you handle the **publishing** (`AttachState.Attaching`) state depends on the situation. If you want to prevent `container.attach` from being called when it has already been called, then treat publishing (`Attaching`) the same as published (`Attached`). 
On the other hand, if you want to block editing when until the container is fully published, treat publishing (`Attaching`) the same as unpublished (`Detached`).

> 

Reviewer: I get the sprit behind this but I think it deserves as a caveat that the user wants to treat `Detached` as "no edits allowed" in the first place, which isn't necessarily the case (above we mention how data in the container can be edited right away while it is detached).

Also, `"editing when until"` has an extra word :).

Synchronization status refers to whether the container's data on the client is saved to the Fluid service. It is a *client-relative state*: the container may have a different synchronization state on different clients. The following diagram shows the possible Synchronization states and the events that cause a state transition. Details are below the diagram.

<!-- TO MODIFY THIS DIAGRAM, SEE INSTRUCTIONS AT THE BOTTOM OF THIS FILE. -->
![A state diagram of the two possible Synchronization states](./images/SynchronizationStates.png)

{{< callout note >}}

The dotted arrow represents a boolean guard condition. It means that the container cannot transition out of the **dirty** state unless the container is published and connected.

{{< /callout >}}

- **dirty**: A container is in **dirty** state on a client in two situations:

    - It is unpublished. 
    - Changes have been made to the data on the client but not yet sent to the Fluid service, or sent but not yet acknowledged by the service.

- **saved**: A container is in **saved** state on a client when the container has been published and the service has acknowledged all data changes made on the client. 

> When a new change is made to the data, the container moves to **dirty** state. When all pending changes are acknowledged, it transitions to the **saved** state. 

Reviewer: ```suggestion
When a new change is made to the data in a given client, the container moves to **dirty** state *in that client*. When all pending changes are acknowledged, it transitions to the **saved** state. 
```
Just to be clear that changes in other clients don't affect the local client's dirty/saved state.

Users can work with the container's data regardless of whether it is in **dirty** or **saved** state. 
But there are scenarios in which your code must be aware of the container's state. For this reason, the `container` object has a boolean `isDirty` property to specify its state. 
The `container` object also supports *dirty* and *saved* events, so you can handle the transitions between states.
The container emits the *saved* event to notify the caller that all the local changes have been acknowledged by the service.

```typescript {linenos=inline}
container.on("saved", () => {
    // All pending edits have been saved.
});
```

The container emits the *dirty* event to notify the caller that there are local changes that have not been acknowledged by the service yet.

```typescript {linenos=inline}
container.on("dirty", () => {
    // The container has pending changes that need to be acknowledged by the service.
});
```

You should always check the `isDirty` flag before [disposing](#disposed) the container or [disconnecting](#disconnected) from the service.

> If you dispose or disconnect the container while `isDirty === true`, you may lose operations that have not yet been acknowledged by the service.

Reviewer: ```suggestion
If you dispose or disconnect the container while `isDirty === true`, you may lose operations that have not yet been sent to (or acknowledged by) the service.
```

An interesting scenario here is that an operation might have been sent to the service, the container disconnects/disposes, but the service still manages to sequences that operation thus making it durable even if the client that sent it didn't get the acknowledgement. I think this is possible in principle. The main scenario where operations can be lost is when they have not been sent to the service at all.

- Your code calls the client's `getContainer` method on a client that has not previously been connected. For more information, see [Connecting to a container](./containers.md#connecting-to-a-container).
- The Fluid client runtime tries to reconnect following a disconnection caused by a network problem.
- Your code calls the `container.connect` method on a client that had become disconnected. 

This process is normally very brief. When it completes, the container automatically transitions to the **catching up** state while changes made on other clients are relayed to the current client. 

#### Catching up

In this state, a partial connection has been established and the client is receiving changes made in other clients from the Fluid service. 

Changes made locally are not uploaded to the service until catching up is complete, so the container is not considered fully connected.

This state is normally very brief and the Fluid client then automatically moves to the **connected** state. 

> When the Fluid service resumes relaying data, the container automatically transitions back to the **catching up** state.

Reviewer: Is this correct? I'm not super familiar with the area but I would have expected that catching up only occurs between **establishing connection** and **connected**. Once connected, the Fluid service is constantly relaying data, but I think a client just stays **connected** going forward.

Author: It was not correct. Fixed now.

#### Connected

A container is **connected** when there is an open, two-way web socket connection between the client and the Fluid service. 
In the **connected** state, changes to the container's data on the client are sent to the service which relays them to all other connected clients. 

The container transitions to this state automatically when it is not receiving relayed data from the Fluid service. 

> When new data is sent from the service, the container automatically transitions to the **catching up** state and then transitions back to **connected** when the catching up is complete.

Author: fixed

```typescript
user.on("idle", () => {
    // Disconnect the container when the user is idle.
    container.disconnect();
});

user.on("active", () => {
    // Connect the container when the user is active again.
    container.connect();
});
```

### Local Readiness state

Local Readiness is a *client-relative state*: the container may have a different Local Readiness state on different clients. The following diagram shows the possible Local Readiness states and the events that cause a state transition. Details are below the diagram.

<!-- TO MODIFY THIS DIAGRAM, SEE INSTRUCTIONS AT THE BOTTOM OF THIS FILE. -->
![A state diagram of the two possible Local Readiness states](./images/LocalReadinessStates.png)

#### Ready

In this state, changes to the shared data objects can be made locally. They are shared with the Fluid service when, and only when, the container is in **published** and **connected** status. 

A container is automatically in **ready** state as soon as it is created and it remains in this state unless it is disposed. 

> A disposed container transitions back to **ready** if its ID is passed to a call of `client.getContainer`.

Reviewer 2: The specific instance of the FluidContainer remains disposed forever - probably clarify that this isn't a transition, but rather just that a new "ready" container can be retrieved if needed.
```suggestion
A disposed `FluidContainer` remains disposed forever, but a new **ready** `FluidContainer` can be created by passing its ID to a call of `client.getContainer`.
```

> A disposed container transitions back to **ready** if its ID is passed to a call of `client.getContainer`.

Author: @alexvy86 Suppose the following code. 

```
const documentId= await container.attach();

container.dispose();

const { newContainer, services } =
    await client.getContainer(documentId schema);

// Assign a ready container to the disposed container.
container = newContainer;
```

Granted that `container` has a different containerId than it used to have, it has the same documentId, so could this not be considered a transition from disposed state to ready state? Or is that too much of a stretch?

> A disposed container transitions back to **ready** if its ID is passed to a call of `client.getContainer`.

Reviewer: I'd say it's too much of a stretch, and/or potentially confusing once users encounter the container/document distinction. When reading this doc I'm thinking in terms of state machines/transitions, especially thanks to the diagrams, so if I read that something can transition from X to Y, I picture that exact object changing states, but in this case we're just replacing the object that the container *variable* points to, with a new one which is in another state that is not reachable anymore by the original object. It does become trickier given that what we mean by "container" in the docs is not the same thing the code means by it, but even trying to put that aside in my mind (not easy 😄), I think it'd be misleading (and potentially counterproductive) to suggest a disposed container can change back. Instead I'd point the users to "if you want to interact with a container after it has been disposed, discard the disposed object and load the container again into a new one".

#### Disposed

In scenarios where a container is no longer needed on the current client, you can dispose it with a call of `container.dispose()`. Disposing removes any server connections, so the Connection status becomes **disconnected**.
There is a *disposed* event on the container object that you can handle to add custom clean up logic, such as removing registered events.
The following shows the basic syntax:

```typescript
container.on("disposed", () => {
    // Handle event cleanup to prevent memory leaks.
});
```

It is a good practice to make sure all pending changes are saved prior to calling `dispose`.
The following shows how to ensure that the disposal doesn't happen until the container is **saved**.

```typescript
function disposeWhenSafe {
    if (container.isDirty) {
        container.once("saved", () => {
            container.dispose();
        });
    }
}
```

> After it is disposed, some properties of the `container` object can be read, but the data objects in the container can no longer be changed either locally of from data relayed by the Fluid service.

Author: > This is work in progress as far as I know, but we're moving towards a distinction between a **closed** container (changes not allowed anymore, data from other clients is not coming in, but reads are possible) and a **disposed** container (neither changes nor reads are allowed anymore). Might become relevant and require some updates here.

I think that will have to be another PR if and when this happens.

Once a container is disposed, your code must call `client.getContainer` again if it needs to be reconnected.

> This also moves it back to the **ready** state.

Reviewer: See a previous comment, I don't think we want to give the impression that a disposed/closed container can go back to the ready state. The underlying document needs to be loaded in a new container.

- **Publication status**: 

    - unpublished
    - publishing (to the Fluid service)
    - published
    - deleted (from the service)

- **Synchronization status**: 

    - dirty (on a given client)
    - saved

- **Connection status**: 

    - disconnected (from the Fluid service)
    - establishing connection (to the service)
    - catching up (with data from other clients)
    - connected

- **Local readiness status**: 

    - ready
    - disposed (on the client)

### Publication status states

Publication status refers to whether or not the container has been *initially* saved to the Fluid service, and whether it is still persisted there. 
Think of it as a mainly *service-relative* state because it is primarily about the container's state in the Fluid service and secondarily about its state on the creating client. The following diagram shows the possible publication states and the events that cause a state transition. Details are below the diagram.

<!-- TO MODIFY THIS DIAGRAM, SEE INSTRUCTIONS AT THE BOTTOM OF THIS FILE. -->

> ![A state diagram of the four possible publication states](./images/PublicationStates.svg)

Reviewer 3: To fix the image links, do the following:

1. Create a new folder called container-states-events under the build folder.
2. Move the container-states-events.md file into that folder and rename it to index.md
3. Move the images folder under the container-states-events folder.

The links should start working after that.
